athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/handlers/MetadataHandler.java

```diff
@@ -240,10 +281,20 @@ protected final void doHandleRequest(BlockAllocator allocator,
                 }
                 return;
             case LIST_TABLES:
-                try (ListTablesResponse response = doListTables(allocator, (ListTablesRequest) req)) {
-                    logger.info("doHandleRequest: response[{}]", response);
-                    assertNotNull(response);
-                    objectMapper.writeValue(outputStream, response);
+                if (((ListTablesRequest) req).getNextToken().isPresent()) {
+                    // Process paginated LIST_TABLES request.
+                    try (ListTablesResponse response = doPaginatedListTables(allocator, (ListTablesRequest) req)) {
+                        logger.info("doHandleRequest (w/pagination): response[{}]", response);
+                        assertNotNull(response);
+                        objectMapper.writeValue(outputStream, response);
+                    }
+                }
+                else {
+                    try (ListTablesResponse response = doListTables(allocator, (ListTablesRequest) req)) {
+                        logger.info("doHandleRequest: response[{}]", response);
+                        assertNotNull(response);
+                        objectMapper.writeValue(outputStream, response);
+                    }
```

Comment:
we shouldnt need to branch here.... its just one thing

Reply:
Pagination implementation moved to the connector.